test/extended/router/gatewayapicontroller.go

```diff
@@ -0,0 +1,502 @@
+			o.Expect(err).NotTo(o.HaveOccurred())
+			e2e.Logf("The gateway Aaddress is: %v", gwwAddress)
+			o.Expect(lbExternalIP).To(o.Equal(gwwAddress))
+		})
+
+		g.It("and ensure HTTPRoute object is created", func() {
+			g.By("Getting the default domain for creating custom gateway")
+			defaultIngressDomain, err := getDefaultIngressClusterDomainName(oc, time.Minute)
+			o.Expect(err).NotTo(o.HaveOccurred(), "failed to find default domain name")
+			defaultDomain := strings.Trim(defaultIngressDomain, "apps.")
+
+			g.By("Create the custom API Gateway")
+			createGateway(oc, customGatewayName, gatewayClassName, defaultDomain)
+
+			g.By("Create the http route using the custom gateway")
+			defaultRoutename := "test-hostname.gwapi." + defaultDomain
+			createHttpRoute(oc, "test-httproute", defaultRoutename, "echo-pod-"+customGatewayName)
+
+			g.By("Checking the http route using the custom gateway is accepted")
+			assertHttpRouteSuccessful(oc, "test-httproute")
+		})
+
+	})
+})
+
+func checkGatewayClass(oc *exutil.CLI, name string) error {
+	gwapiClient := gatewayapiclientset.NewForConfigOrDie(oc.AdminConfig())
+
+	waitErr := wait.PollUntilContextTimeout(context.Background(), 2*time.Second, 2*time.Minute, false, func(context context.Context) (bool, error) {
+		gwc, err := gwapiClient.GatewayV1().GatewayClasses().Get(context, name, metav1.GetOptions{})
+		if err != nil {
+			e2e.Logf("failed to get gatewayclass %s, retrying...", name)
+			return false, nil
+		}
+		for _, condition := range gwc.Status.Conditions {
+			if condition.Type == string(gatewayapiv1.GatewayClassConditionStatusAccepted) {
+				if condition.Status == metav1.ConditionTrue {
+					return true, nil
+				}
+			}
+		}
+		e2e.Logf("Found gatewayclass %s but it is not accepted, retrying...", name)
+		return false, nil
+	})
+
+	if waitErr != nil {
+		return fmt.Errorf("Gatewayclass %s is not accepted", name)
+	}
+	e2e.Logf("Gateway Class %s is created and accepted", name)
+	return nil
+}
+
+// buildGatewayClass initializes the GatewayClass and returns its address.
+func buildGatewayClass(name, controllerName string) *gatewayapiv1.GatewayClass {
+	return &gatewayapiv1.GatewayClass{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+		Spec: gatewayapiv1.GatewayClassSpec{
+			ControllerName: gatewayapiv1.GatewayController(controllerName),
+		},
+	}
+}
+
+// createGateway build and creates the Gateway.
+func createGateway(oc *exutil.CLI, gwname, gwclassname, domain string) *gatewayapiv1.Gateway {
+	gwapiClient := gatewayapiclientset.NewForConfigOrDie(oc.AdminConfig())
+	ingressNameSpace := "openshift-ingress"
+	gateway := &gatewayapiv1.Gateway{}
+
+	// Get getway class details to create gateway
+	gatewayClass, errGwClass := gwapiClient.GatewayV1().GatewayClasses().Get(context.TODO(), gwclassname, metav1.GetOptions{})
+	if errGwClass != nil {
+		e2e.Failf("Expected gateway class object but not found, the error is %v", errGwClass)
+	}
```

Comment:
Can we just pass the gatewayclass name to the function to avoid getting it from OpenShift again?